docs/ref/models/fields.txt

```diff
@@ -115,9 +115,33 @@ human-readable name. For example::
         ("GR", "Graduate"),
     ]
 
+``choices`` can also be defined as a callable that expects no arguments and
+returns any of the formats described above. This could be handy when generating
+choices that may involve expensive calculations, or performing time-consuming
+I/O-bound operations. For example::
+
+    import functools
+
+
+    @functools.cache
+    def fib(n):
+        if n < 2:
+            return n
+        return fib(n - 1) + fib(n - 2)
+
+
+    @functools.cache
```

Comment:
It might be worth checking that `@functools.cache` works nicely in the migrations serializer by adding a test...

I think that, as long as decorators use `functools.wraps()` properly we should get the right output from `FunctionTypeSerializer`, but probably worth checking.

Reply:
Thank you for raising this. As I think you suspected, the migrations serializer fails with

```
ValueError: Cannot serialize: <functools._lru_cache_wrapper object at 0x7f819ab3b270>
```

I've checked and the Python code does not wraps the function, so I created [a topic in the Python discourse](https://discuss.python.org/t/why-functools-cache-or-lru-cache-does-not-seem-to-wraps-the-passed-function/32965) (though it was held for moderation :eyeglasses:). Depending on what the Python devs say, this behavior may be a bug or not, and we may need to workaround it in our migration serializer.

Reply:
Turns out we may need to update the migrations serializer to either:
  1. Provide a specific serializer for the `_lru_cache_wrapper` type, or
  2. Provide a serializer for anything that has a `__wrapped__` attribute

Do you have an opinion on the above? Whatever we do, it may need a discussion in the forum and likely a separated ticket/PR.

Reply:
I created [ticket 34822](https://code.djangoproject.com/ticket/34822) to provide support for the `_lru_cache_wrapper`, which I think, at this stage, will be a 5.1 feature (if accepted).